python/ray/autoscaler/v2/instance_manager/subscribers/instance_launcher.py

```diff
@@ -0,0 +1,162 @@
+                    ],
+                )
+
+    def _launch_new_instances_by_type(
+        self, instance_type: str, instances: List[Instance]
+    ) -> int:
+        """Launches instances of the given type.
+
+        Args:
+            instance_type: type of instance to launch.
+            instances: list of instances to launch. These instances should
+                have been marked as QUEUED with instance_type set.
+        Returns:
+            num of instances launched.
+        """
+        logger.info(f"Launching {len(instances)} instances of type {instance_type}")
+        instances_selected = []
+        for instance in instances:
+            instance.status = Instance.REQUESTED
+            result, version = self._instance_storage.upsert_instance(
+                instance, expected_instance_version=instance.version
+            )
+            if not result:
+                logger.warn(f"Failed to update instance {instance}")
+                continue
+            instance.version = version
+            instances_selected.append(instance)
+
+        if not instances_selected:
+            return 0
+
+        # TODO: idempotency token.
+        created_cloud_instances = self._node_provider.create_nodes(
+            instance_type, len(instances_selected)
+        )
+
+        assert len(created_cloud_instances) <= len(instances_selected)
+
+        instances_launched = 0
+        while created_cloud_instances and instances_selected:
+            cloud_instance = created_cloud_instances.pop()
+            instance = instances_selected.pop()
+            instance.cloud_instance_id = cloud_instance.cloud_instance_id
+            instance.internal_ip = cloud_instance.internal_ip
+            instance.external_ip = cloud_instance.external_ip
+            instance.status = Instance.ALLOCATED
+            instance.ray_status = Instance.RAY_STATUS_UNKOWN
+
+            # update instance status into the storage
+            result, _ = self._instance_storage.upsert_instance(
+                instance, expected_instance_version=instance.version
+            )
+
+            if not result:
+                # TODO: this could only happen when the request is canceled.
+                logger.warn(f"Failed to update instance {instance}")
+                # push the cloud instance back
+                created_cloud_instances.append(cloud_instance)
+                continue
+
+            instances_launched += 1
+
+        if created_cloud_instances:
+            # instances are leaked, we probably need to terminate them
+            for instance in created_cloud_instances:
+                self._node_provider.terminate_node(instance.cloud_instance_id)
+
+        if instances_selected:
+            # instances creation failed, we need to marke them allocation failed.
+            for instance in instances_selected:
+                instance.status = Instance.ALLOCATION_FAILED
+                # TODO: add more information about the failure.
```

Comment:
Great details, I'm interested in failure information. Please let me know if it's already in the pipeline.

1. Is this failure reason more about node_provider launch failure? If no, can we add that information? For example an EC2 instance for worker_node_1 failed to launch with capacity exception.
2. Do we have plans to support as API (retrieve instance_storage)? For example, `autoscaler_v2.get_failed_instances_info(instance_ids: List[str])`

Reply:
re 1: Yes, I think this is when less than selected instances could be started by the node_provider. As for surfacing the specific error info here, maybe we should do so via either metrics/events. Or as you mentioned in 2, through some of APIs to retrieve it. 

cc @scv119